src/ini.cr

```diff
@@ -1,22 +1,59 @@
 class INI
+  # Exception thrown on an INI parse error.
+  class ParseException < Exception
+    getter line_number : Int32
+    getter column_number : Int32
+
+    def initialize(message, @line_number, @column_number)
+      super "#{message} at #{@line_number}:#{@column_number}"
+    end
+
+    def location
+      {line_number, column_number}
+    end
+  end
+
   # Parses INI-style configuration from the given string.
+  # Raises a `ParseException` on any errors.
   #
   # ```
   # INI.parse("[foo]\na = 1") # => {"foo" => {"a" => "1"}}
   # ```
   def self.parse(str) : Hash(String, Hash(String, String))
     ini = {} of String => Hash(String, String)
+    current_section_name = ""
+    lineno = 0
 
-    section = ""
     str.each_line do |line|
-      if line =~ /\s*(.*[^\s])\s*=\s*(.*[^\s])/
-        ini[section] ||= {} of String => String if section == ""
-        ini[section][$1] = $2
-      elsif line =~ /\[(.*)\]/
-        section = $1
-        ini[section] = {} of String => String
+      lineno += 1
+      next if line.empty?
+
+      offset = 0
+      line.each_char do |char|
+        break unless char.ascii_whitespace?
+        offset += 1
+      end
+
+      case line[offset]
+      when '#', ';'
+        next
+      when '['
+        end_idx = line.index(']', offset)
+        raise ParseException.new("unterminated section", lineno, line.size) unless end_idx
+        raise ParseException.new("data after section", lineno, end_idx + 1) unless end_idx == line.size - 1
+
+        current_section_name = line[1...end_idx]
+        ini[current_section_name] = {} of String => String
+      else
+        key, eq, value = line.partition('=')
+        raise ParseException.new("expected declaration", lineno, key.size) if eq != "="
+
+        section = ini[current_section_name]? || Hash(String, String).new
```

Comment:
We can just store the `current_section_hash` and then we only have to do a single hash op for each k=v, instead of 3 hash operations.

```cr
ini = Hash(String, Hash(String, String)).new
current_section_name = ""
current_section = ini[current_section_name] = Hash(String, String).new
```

then in `[` just
```cr
current_section_name = line[1...end_idx]
current_section = ini[current_section_name] = Hash(String, String).new
```
and that simplifies `else` to just
```cr
current_section[key.strip] = value.strip
```

 and you're done!

Reply:
Great idea.

This doesn't need `current_section_name` as a static variable outside the loop.

```crystal
ini = Hash(String, Hash(String, String)).new
current_section = ini[""] = Hash(String, String).new
```

Reply:
I went with

```crystal
current_section = ini[current_section_name] ||= Hash(String, String).new
```

To allow for reopened sections, since the original parser allows them (but didn't test for them in the spec).

Reply:
Good catch